### ISSUE TITLE
Dogfood

### ISSUE THREAD

**Issue by sageserpent-open** [sageserpent-open/curium#2]
`Curium` needs to be used in a real project - tests and benchmarks are all well and good, but there's nothing like grappling with something that needs to be shipped to bring out the pain points in an API / highlight missing capabilities / flush out exciting and subtle bugs.

Yes, the `Plutonium` project has served that purpose in the past, but as it is on hold for a while, something pithier is needed.

**Comment by sageserpent-open** [comment]
A service with REST endpoints that allow a data model to be updated and queried, using `Curium` to store the data model.

Excitement is furnished by implementing this with a view to hosting it on AWS, allowing scaling and failover across an EC2 instance auto-scaling group / EKS / ECS or whatever.

Current possibilities are:

1. Use the  RocksDB tranches implementation, so that the leader service process writes to a file system in either the root volume of an EC2 instance or to an additional instance volume. Use RocksDb to replicate the store to EFS, and then replicate back to the secondary service process instances to serve queries / be ready to acquire leadership.
2. Provide a tranches implementation on top of DynamoDB, or maybe S3.
3. Port the existing H2 tranches implementation over to use Aurora or some other RDS implementation instead, and share that between the EC2 instances.

**Comment by sageserpent-open** [comment]
`Curium` works best with updates batch together in a single session, so the web framework needs to be able to batch lots of web requests and then respond to them once the entire batch is done - so *not* the Spring controller model of 'call a method to make changes for a single request and then return a single response like an RPC call'.

Can Akka do that?

**Comment by sageserpent-open** [comment]
Well, I started out on this and wrote something that could be plugged into Play to service requests on a non-trivial data model (this is in its own private repository, not in this one). There is a batching framework, that was tested and debugged in isolation, then the data model and Curium were plugged in to provide the service's core logic.

A whole lot of benchmarking directly against the resulting core logic was done to see whether this approach would scale - and a whole lot of bugs were found and fixed. So far, so good.

Nevertheless, performance of the benchmark is ropey - yes, the model is updated, and yes, it makes steady progress and fits in about 16G of memory, and it stays up and running without falling over - but every now and then, batches stutter with long delays of >= 16 seconds for an individual batch to complete. This maximum latency just isn't acceptable, although the rest of the time batches come through in < 1 seconds for a batch of 5 to 20 operations.

So it works, sort of, but it isn't a contender.

**Comment by sageserpent-open** [comment]
The aforementioned bugs have had tests written to reproduce them outside of the private repository code, and these have been fixed and merged back on to `master`.

**Comment by sageserpent-open** [comment]
There is one last thing to check - I'm not completely sure whether the original benchmark of `ImmutableObjectMeetsMap` is also performing with a stutter. If it *isn't*, then OK, either it isn't quite realistic enough or the benchmark for the service's core logic may itself be suspect. If it is stuttering, it's worth revisiting the benchmark against older commits to see if a performance regression has been introduced...

**Comment by sageserpent-open** [comment]
After some fettling of `ImmutableObjectMeetsMap`, it appears not to stutter that much - when doing batched updates of 100 at a time, the slowest batches go take around 4 to 5 seconds, the average batch time is around 500 milliseconds. That is for _100_ updates per batch, so could be considered acceptable.

Choosing a smaller batch size has been observed to yield less spiky behaviour at the cost of somewhat lower overall throughput.

An insight gleaned was a) to be very careful that queries make efficient slices into the data inside the `Session` monad and do not try to unpack a giant object _outside_ that monad and that b) it is worth keeping separate intersession state instances for queries versus updates. This might bear fruit in the other repository's benchmark too...

**Comment by sageserpent-open** [comment]
[info] Step: 2510000, duration to before query: 10340433 milliseconds, query duration: 9 milliseconds, minimum update duration: 11 milliseconds, maximum update duration: 715 milliseconds, average update duration: 24, update durations: List((1,715), (1,701), (1,616), (1,615), (1,58), (1,54), (1,43), (1,11), (2,12), (3,38), (3,37), (3,36), (3,34), (5,35), (5,33), (6,30), (8,32), (9,13), (10,31), (17,14), (20,28), (25,16), (28,27), (30,15), (31,29), (44,26), (62,17), (67,25), (73,24), (77,18), (87,22), (89,19), (90,21), (97,23), (98,20)) ,contents at previous step: HashSet(7285, 14729, 226770, 188583, 2509999, 1778695)
[info] Step: 2515000, duration to before query: 10364472 milliseconds, query duration: 9 milliseconds, minimum update duration: 10 milliseconds, maximum update duration: 699 milliseconds, average update duration: 23, update durations: List((1,699), (1,691), (1,425), (1,230), (1,66), (1,62), (1,38), (2,42), (2,35), (2,34), (2,10), (3,32), (5,33), (5,13), (5,12), (11,31), (13,30), (17,28), (18,29), (18,14), (25,15), (34,27), (44,25), (44,16), (56,17), (62,26), (70,24), (82,22), (84,18), (85,19), (88,23), (104,21), (112,20)) ,contents at previous step: HashSet(52628, 420006, 2514999, 32199, 45485, 392392)
[info] Step: 2520000, duration to before query: 10389146 milliseconds, query duration: 9 milliseconds, minimum update duration: 9 milliseconds, maximum update duration: 662 milliseconds, average update duration: 24, update durations: List((1,662), (1,644), (1,626), (1,619), (1,46), (1,45), (1,38), (1,36), (1,11), (1,9), (3,37), (3,34), (3,12), (4,35), (7,33), (10,32), (11,31), (12,13), (13,29), (16,30), (19,14), (22,28), (31,15), (40,16), (41,26), (42,27), (54,17), (61,25), (73,24), (74,23), (76,19), (82,21), (89,18), (97,22), (107,20)) ,contents at previous step: HashSet(6145, 805768, 5984, 2519999, 33975, 24802, 884288, 13629)
[info] Step: 2525000, duration to before query: 10414400 milliseconds, query duration: 8 milliseconds, minimum update duration: 11 milliseconds, maximum update duration: 696 milliseconds, average update duration: 24, update durations: List((1,696), (1,636), (1,571), (1,482), (1,240), (1,45), (1,43), (1,40), (1,12), (2,48), (2,39), (2,38), (2,37), (2,11), (3,36), (4,35), (6,32), (6,13), (9,34), (10,14), (11,33), (11,31), (13,30), (18,29), (20,15), (27,28), (32,16), (35,27), (48,26), (50,17), (53,25), (69,24), (71,18), (88,19), (91,23), (100,21), (101,20), (105,22)) ,contents at previous step: HashSet(2807, 2524999, 6828, 84, 1241985, 572444, 955033)
[info] Step: 2530000, duration to before query: 10439523 milliseconds, query duration: 11 milliseconds, minimum update duration: 9 milliseconds, maximum update duration: 699 milliseconds, average update duration: 24, update durations: List((1,56), (1,49), (1,48), (1,38), (1,37), (1,9), (2,12), (2,11), (3,699), (3,36), (4,35), (4,13), (5,32), (6,14), (7,34), (10,33), (13,31), (18,30), (24,15), (32,16), (34,29), (34,28), (44,27), (46,26), (49,17), (59,18), (75,19), (77,25), (78,24), (81,20), (86,23), (93,21), (105,22)) ,contents at previous step: HashSet(98, 1569549, 801, 1118978, 6, 228290, 239, 989015, 12039, 2529999, 61063, 19171, 1065159, 2, 7872)
[info] Step: 2535000, duration to before query: 10464988 milliseconds, query duration: 10 milliseconds, minimum update duration: 11 milliseconds, maximum update duration: 704 milliseconds, average update duration: 24, update durations: List((1,704), (1,668), (1,629), (1,613), (1,55), (1,52), (1,41), (1,40), (1,11), (3,37), (3,34), (4,33), (4,14), (6,36), (6,35), (8,32), (9,13), (16,31), (16,15), (17,29), (23,30), (25,28), (31,16), (38,27), (45,26), (59,17), (65,25), (65,18), (78,24), (83,19), (84,20), (95,21), (103,23), (105,22)) ,contents at previous step: HashSet(2020, 550000, 2534999, 1892330, 17003, 86618, 964, 47574, 24675, 1048227, 2383)
[info] Step: 2540000, duration to before query: 10489626 milliseconds, query duration: 8 milliseconds, minimum update duration: 7 milliseconds, maximum update duration: 684 milliseconds, average update duration: 24, update durations: List((1,684), (1,596), (1,478), (1,241), (1,39), (1,38), (1,12), (1,11), (1,7), (2,46), (4,36), (4,35), (4,33), (6,34), (6,13), (7,14), (8,31), (10,32), (10,30), (19,29), (19,15), (32,28), (38,16), (43,27), (44,17), (46,26), (65,25), (74,18), (78,24), (79,19), (88,20), (101,23), (101,21), (103,22)) ,contents at previous step: Set(1345239, 2539999)
[info] Step: 2545000, duration to before query: 10515583 milliseconds, query duration: 10 milliseconds, minimum update duration: 12 milliseconds, maximum update duration: 874 milliseconds, average update duration: 25, update durations: List((1,874), (1,687), (1,521), (1,323), (1,282), (1,215), (1,73), (1,56), (2,43), (2,12), (3,34), (3,13), (4,37), (4,36), (5,35), (8,33), (10,31), (13,14), (14,32), (24,30), (29,15), (34,16), (37,27), (38,29), (43,28), (46,17), (47,26), (63,25), (63,18), (74,20), (79,19), (81,23), (86,24), (89,22), (91,21)) ,contents at previous step: HashSet(1740, 1957503, 865, 14566, 2544999, 45144, 949809, 80721)
[info] Step: 2550000, duration to before query: 10541420 milliseconds, query duration: 8 milliseconds, minimum update duration: 11 milliseconds, maximum update duration: 742 milliseconds, average update duration: 25, update durations: List((1,742), (1,589), (1,529), (1,433), (1,397), (1,245), (1,42), (1,41), (1,12), (2,11), (3,37), (3,36), (3,32), (4,38), (4,35), (4,34), (5,14), (5,13), (9,33), (10,31), (15,15), (20,30), (23,29), (24,28), (24,16), (46,27), (49,17), (61,26), (65,18), (75,25), (76,19), (78,24), (89,23), (90,21), (94,22), (110,20)) ,contents at previous step: Set(39603, 5231, 599085, 914693)
[info] Step: 2555000, duration to before query: 10565936 milliseconds, query duration: 9 milliseconds, minimum update duration: 12 milliseconds, maximum update duration: 651 milliseconds, average update duration: 24, update durations: List((1,651), (1,613), (1,530), (1,468), (1,277), (1,49), (1,43), (1,42), (2,46), (2,37), (2,34), (3,36), (4,35), (4,33), (5,12), (8,32), (10,13), (12,30), (14,31), (16,14), (21,29), (21,28), (31,26), (32,15), (36,27), (46,16), (50,25), (55,24), (66,17), (75,18), (87,22), (91,19), (92,20), (94,23), (113,21)) ,contents at previous step: HashSet(1576339, 2554999, 19917, 411438, 29289)
[info] Step: 2560000, duration to before query: 10590481 milliseconds, query duration: 8 milliseconds, minimum update duration: 10 milliseconds, maximum update duration: 671 milliseconds, average update duration: 24, update durations: List((1,671), (1,660), (1,474), (1,239), (1,44), (1,40), (1,38), (1,37), (1,35), (1,11), (1,10), (2,50), (2,43), (2,36), (4,12), (5,34), (5,33), (8,32), (9,13), (11,31), (13,30), (14,14), (19,29), (19,28), (19,15), (30,16), (39,27), (48,17), (53,25), (56,26), (61,18), (75,24), (79,19), (91,20), (95,22), (108,23), (122,21)) ,contents at previous step: HashSet(79561, 75148, 28795, 88185, 96071, 81229, 2559999)
[info] Step: 2565000, duration to before query: 10615600 milliseconds, query duration: 8 milliseconds, minimum update duration: 9 milliseconds, maximum update duration: 721 milliseconds, average update duration: 24, update durations: List((1,721), (1,698), (1,694), (1,648), (1,59), (1,56), (1,54), (1,48), (1,41), (1,39), (1,38), (1,11), (1,9), (2,40), (2,36), (2,33), (2,12), (2,10), (3,37), (4,31), (4,13), (5,34), (7,35), (7,32), (14,28), (14,14), (19,29), (21,30), (26,15), (38,27), (38,26), (51,17), (52,16), (60,25), (69,18), (72,24), (72,19), (83,23), (100,21), (108,22), (110,20)) ,contents at previous step: Set(1277838, 1385115, 2564999)
[info] Step: 2570000, duration to before query: 10639821 milliseconds, query duration: 7 milliseconds, minimum update duration: 10 milliseconds, maximum update duration: 713 milliseconds, average update duration: 23, update durations: List((1,713), (1,509), (1,411), (1,242), (1,241), (1,57), (1,50), (1,39), (1,38), (1,10), (2,37), (2,36), (2,35), (2,12), (3,34), (4,33), (5,31), (9,13), (10,32), (11,29), (12,30), (12,14), (15,28), (22,15), (31,16), (35,27), (49,25), (59,26), (61,17), (83,23), (87,18), (90,22), (91,24), (96,21), (98,20), (99,19)) ,contents at previous step: Set(2569999, 948389)
[info] Step: 2575000, duration to before query: 10665078 milliseconds, query duration: 6 milliseconds, minimum update duration: 10 milliseconds, maximum update duration: 774 milliseconds, average update duration: 24, update durations: List((1,774), (1,733), (1,631), (1,584), (1,57), (1,55), (1,44), (1,40), (1,12), (1,10), (2,39), (2,11), (3,36), (5,37), (5,35), (5,33), (7,34), (8,32), (10,13), (11,14), (13,31), (18,30), (22,29), (23,15), (26,28), (39,26), (40,16), (46,27), (51,17), (54,25), (61,24), (76,18), (88,21), (89,23), (92,22), (94,19), (100,20)) ,contents at previous step: Set(2574999)
[info] Step: 2580000, duration to before query: 10722704 milliseconds, query duration: 5 milliseconds, minimum update duration: 9 milliseconds, maximum update duration: 14203 milliseconds, average update duration: 57, update durations: List((1,14203), (1,2524), (1,823), (1,553), (1,538), (1,389), (1,381), (1,184), (1,126), (1,105), (1,104), (1,91), (1,82), (1,73), (1,66), (1,64), (1,63), (1,9), (2,65), (2,62), (2,60), (2,15), (2,13), (2,12), (3,61), (3,54), (4,59), (4,57), (6,14), (7,25), (8,56), (9,22), (11,53), (11,28), (11,27), (11,24), (11,18), (12,16), (13,55), (13,29), (13,17), (15,52), (16,23), (17,51), (17,21), (17,19), (18,30), (18,26), (19,50), (21,20), (23,33), (24,49), (24,47), (24,31), (26,48), (28,46), (28,45), (30,32), (33,34), (37,44), (37,35), (38,37), (39,43), (40,40), (44,41), (44,38), (45,42), (48,39), (50,36)) ,contents at previous step: HashSet(2579999, 1529818, 1138465, 877366, 1252363)
[info] Step: 2585000, duration to before query: 10767172 milliseconds, query duration: 8 milliseconds, minimum update duration: 23 milliseconds, maximum update duration: 658 milliseconds, average update duration: 43, update durations: List((1,658), (1,592), (1,569), (1,550), (1,533), (1,433), (1,345), (1,141), (1,129), (1,121), (1,88), (1,68), (1,67), (1,64), (1,26), (1,24), (2,60), (2,59), (2,58), (2,23), (3,57), (3,55), (3,28), (4,25), (5,52), (5,27), (6,56), (6,53), (8,29), (12,54), (13,50), (13,30), (14,51), (18,49), (29,48), (29,31), (30,32), (31,47), (31,33), (42,45), (44,46), (44,34), (48,43), (49,44), (56,41), (57,38), (58,37), (59,40), (61,35), (62,42), (63,36), (70,39)) ,contents at previous step: Set(2584999)
[info] Step: 2590000, duration to before query: 10808831 milliseconds, query duration: 6 milliseconds, minimum update duration: 24 milliseconds, maximum update duration: 545 milliseconds, average update duration: 41, update durations: List((1,545), (1,536), (1,521), (1,502), (1,431), (1,239), (1,119), (1,112), (1,63), (1,62), (1,60), (2,118), (2,24), (3,54), (3,53), (3,25), (5,49), (5,26), (7,29), (8,52), (8,28), (8,27), (10,50), (10,47), (11,51), (18,48), (29,30), (30,46), (33,45), (33,31), (35,44), (46,32), (47,43), (51,42), (53,34), (56,41), (63,38), (63,37), (63,33), (65,40), (65,39), (76,36), (78,35)) ,contents at previous step: Set(1813245, 2589999)
[info] Step: 2595000, duration to before query: 10849231 milliseconds, query duration: 6 milliseconds, minimum update duration: 21 milliseconds, maximum update duration: 560 milliseconds, average update duration: 39, update durations: List((1,560), (1,552), (1,542), (1,535), (1,362), (1,229), (1,123), (1,121), (1,110), (1,109), (1,58), (1,53), (1,21), (2,56), (2,55), (3,51), (3,23), (4,54), (4,52), (5,24), (7,50), (8,49), (8,26), (10,25), (11,48), (12,28), (13,47), (15,27), (16,46), (21,29), (24,45), (27,44), (28,43), (34,42), (36,30), (44,31), (50,32), (52,41), (61,40), (62,36), (65,37), (65,33), (67,34), (72,39), (72,35), (84,38)) ,contents at previous step: Set(2594999, 853141)
[info] Step: 2600000, duration to before query: 10889469 milliseconds, query duration: 7 milliseconds, minimum update duration: 17 milliseconds, maximum update duration: 577 milliseconds, average update duration: 39, update durations: List((1,577), (1,568), (1,548), (1,523), (1,356), (1,238), (1,183), (1,137), (1,131), (1,130), (1,128), (1,125), (1,124), (1,76), (1,62), (1,55), (1,24), (1,17), (2,111), (2,49), (2,19), (3,53), (4,56), (4,52), (5,48), (7,51), (7,23), (8,50), (8,25), (11,46), (11,26), (12,47), (14,28), (19,45), (23,27), (24,44), (27,43), (32,42), (32,29), (36,30), (41,41), (47,40), (49,31), (57,39), (58,32), (64,38), (67,34), (72,37), (77,36), (77,35), (80,33)) ,contents at previous step: Set(2599999, 3400)
[info] Step: 2605000, duration to before query: 10926949 milliseconds, query duration: 8 milliseconds, minimum update duration: 19 milliseconds, maximum update duration: 572 milliseconds, average update duration: 36, update durations: List((1,572), (1,552), (1,507), (1,416), (1,389), (1,235), (1,232), (1,113), (1,100), (1,98), (1,53), (1,51), (1,21), (1,20), (1,19), (2,22), (3,50), (3,48), (6,49), (6,47), (9,45), (11,24), (11,23), (12,46), (13,44), (16,42), (17,25), (20,43), (24,26), (27,41), (32,27), (36,40), (41,28), (42,39), (47,38), (50,29), (51,37), (60,32), (60,31), (67,34), (77,36), (77,30), (78,35), (87,33)) ,contents at previous step: HashSet(89562, 658, 410988, 911, 91)
[info] Step: 2610000, duration to before query: 10963657 milliseconds, query duration: 10 milliseconds, minimum update duration: 18 milliseconds, maximum update duration: 565 milliseconds, average update duration: 36, update durations: List((1,565), (1,547), (1,545), (1,532), (1,371), (1,235), (1,108), (1,104), (1,92), (1,57), (1,56), (1,54), (1,19), (1,18), (2,99), (2,51), (3,50), (3,48), (3,20), (4,49), (4,46), (4,22), (6,21), (8,47), (8,23), (9,43), (10,45), (10,24), (14,44), (14,42), (19,41), (22,25), (23,40), (32,26), (35,38), (36,39), (43,27), (47,28), (56,35), (58,37), (61,36), (67,30), (70,29), (71,34), (79,33), (80,31), (83,32)) ,contents at previous step: HashSet(11194, 2030224, 10095, 354065, 45957, 226, 125437, 415, 23213, 410962, 2609999, 48454, 1759, 59373, 502)

**Comment by sageserpent-open** [comment]
What we see above from running `ImmutableObjectMeetsMap` are query times hovering around 7 to 10 milliseconds, and average update times (for batches of 5 operations) hovering around 20 to 60 milliseconds - but with whopping great spikes of 14+ seconds every now and then - these correspond to some kind of GC activity which seems to be related to the tranches cache being emptied every 40 minutes or so.

So, yes it works, but not with strong enough guarantees on latency.

I'm minded to close this ticket as a failed experiment for now, and try an alternative technology for the example project to get a sense of performance using a different approach.

**Comment by sageserpent-open** [comment]
As a footnote, the example project was cutover to use an deferred imperative model using the Doobie `ConnectionIO` monad and some runs were performed using both approaches.

The example project used three parallel streams of requests that were executed by a single-threaded batch processor in batch sizes of three (to match the number of parallel stream inputs).

Verification was done before running the performance tests with a single input stream and batch size of one to prove that both approaches resulted in the same sequence of requests and outcomes.

The example's computational workload becomes increasingly demanding to start with, then passes a maximum and settles down to a constant rate after an hour or so.

**Comment by sageserpent-open** [comment]
With a pure functional model and Curium, after 1 hour 11 minutes and 31 seconds, 164 525 requests were executed. Most batches took from a few tens of milliseconds up to less than 300 milliseconds to complete, but there were spikes from anywhere from 1 second to a whopping 28 seconds every now and then!

Re-running, after 4 hours, 13 minutes and 32 seconds, 778 728 requests were executed. By the end of this, most batches took from a few tens of milliseconds to less than 100 milliseconds to complete, but there were spikes from 1 to 3 seconds.

CPU usage fluctuated rapidly, and memory was consumed in regular sawtooth spikes of around 1G extra per spike.

14G heap was allocated, and the example could not run effectively with much less heap size.

**Comment by sageserpent-open** [comment]
With a deferred imperative model and Doobie, after 1 hour, 11 minutes and 31 seconds, 116 650 requests were executed. Most batches took from a few tens of milliseconds to less than 800 milliseconds to complete, but there were moderate spikes of a little over 1 second every now and then.

Re-running, after 4 hours, 13 minutes and 31 seconds, 839 022 requests were executed. By the end of this, most batches took from a few tens of milliseconds to less than 100 milliseconds to complete, but there were spikes of up to 300 milliseconds.

CPU usage was steady, as was memory usage.

Only 800 Mb of heap was taken by the process (the heap size was not preallocated, rather the JVM defaults were used and the heap allowed to grow on demand from the process).

**Comment by sageserpent-open** [comment]
So we have a tortoise and hare situation here the Curium version starts out faster when the example is building up its computational workload up to and past the maximum, but starts to lag behind the Doobie version once the workload settles down.

The Curium version needs a *lot* of heap - which is not surprising given that the model is essentially a lazy-loaded in-memory one. It also causes a lots of CPU usage fluctuation, and exhibits a regular dramatic sawtooth of memory usage that culminates in dramatic spikes in the batch processing time - these correlated with the sudden drops of memory usage; presumably Curium was having to reload purged entries that were flushed from cache.

The spikes became less dramatic as the workload settled, but never went away completely.

The Doobie version started out with slower performance in the main, but its spikes were far less dramatic to start with. Again, they became even less dramatic as the workload settled, to the point where they could be tolerated.

Performance improved and overtook the Curium version in the long run.

The Doobie version was very economical on heap, fitting comfortably in 800Mb.

**Comment by sageserpent-open** [comment]
Now, it may be the case that some juggling of Curium's caching can be done that minimises the dramatic sawtooth profile (which was also seen in `ImmutableObjectMeetsMap`), but it is certainly clear that the Doobie version was a good, predictable and economical performer.

There is an additional wrinkle here - the Curium version can lazily load pieces of the model data on demand, whereas the Doobie version is currently somewhat naively implemented and loads entire chunks of model data, regardless of whether it is used completely. This definitely has a bearing as the example code works towards the maximum workload - the model data gets a lot larger than what is actually needed by the example code's results, so in that initial phase, the Doobie version has more data to load. This may explain the slower start it exhibited.

**Comment by sageserpent-open** [comment]
Further investigation and consequent code changes have yielded some interesting results, hence re-opening this ticket...

**Comment by sageserpent-open** [comment]
A couple of insights cleaned from recent work on the benchmark:

1. There is state capture from proxies that are cached - each proxy's acquired state is defined within a session interpreter object, so these hang around with the proxies due to the implied reference. It is all very well trying to encapsulate things in a nice clean session interpreter, but this does increase the object bloat.
2. The reference resolvers (whose classes are also local to the session interpreter) were being cached as part of the completed operation data for tranches, so as to perform resolution of requests by proxies to load the underlying objects. These resolvers also capture a reference to the session interpreter, and carry some extra bloat in their own right.
3. The cache of completed operations data is necessary, but isn't the best way of resolving requests for underlying objects  as these can cut across a swathe of tranches - so keeping all of the reference resolvers around for the large number of cached tranches leads to a lot of memory overhead. It is better to slim down the tranches cache, and add a secondary level weak values cache for the resolved underlying objects - thus accommodating requests for 'hot' objects without pulling in entire tranches.
4. Once the decision to allow both the tranches cache to be small (and thus with high turnover) and to use a secondary level cache for underlying objects that itself allows turnover, we can relax a previous important decision, namely to never cache objects that are produced directly by the client code session. The rationale for this used to be that caching such objects would hold on to state that is not made up out of proxied references - so the cached objects would be 'heavy'. Instead, caching of objects was done only on tranche reload, so these objects would be 'light', using proxied references to refer to their internal state. Now that the caching lets go of the objects more frequently, this allows both approaches to be mixed without filling up memory with ever larger objects - in fact, the objects from a client session now consist of a mixture of heavy and light state.
5. There are a lot of closures holding on to odd bits of state that hang around waiting to be garbage collected. They aren't cached, but they add overhead to an already overworked garbage collector.
6. The big pauses are caused ultimately by garbage collection cycles that are in turn being driven by cache turnover. Applying the previous changes reduces the burden on the garbage collector, and this in turn allows the use of the new Z garbage collector. By itself, using the Z garbage collector does not help at all, but in concert with the aforementioned changes it makes a huge difference.


So currently there is an experimental branch in the benchmarking repository with these changes brutally hacked in. Working on this also reveals some design botches in the API that need rectifying - it is now clear that there is no point in passing both the tranches and the intersession state to successive calls on `ImmutableObjectStorage`. Instead, `ImmutableObjectStorage` should layer itself around intersession state and that should in turn later itself around the tranches.

This would allow us to drop the session object, moving the interpretation directly into `ImmutableObjectStorage`.

**Comment by sageserpent-open** [comment]
Some results from benchmarking with the hacked code...

Increasing the parallelism in the benchmark to 20, so having 20 requests in a batch:

Imperative approach using Doobie, Z GC:

2 hours, 23 minutes, 43 seconds --> 221 446 requests

Memory used fluctuates continually between 1.7Gb and 3 Gb, with occasional spikes up to 3.7Gb and troughs down to 1.2Gb.

CPU usage fluctuates between 6% and 10%.

Batches often take >~ 1 second, but settle down to less than once every 5794 requests (290 batches).


|Milliseconds|Number of Requests|
|---|---|
|1 to 100|52|
|100 to 200|1191|
|200 to 300|619|
|300 to 400|1398|
|400 to 500|895|
|500 to 600|463|
|600 to 700|185|
|700 to 800|34|
|800 to 900|9|
|900 to 1000|0|

**Comment by sageserpent-open** [comment]
Imperative approach using Doobie, Default Java 14 GC:

2 hours, 23 minutes, 43 seconds --> 232 942 requests


|Milliseconds|Number of Requests|
|---|---|
|1 to 100|42|
|100 to 200|795|
|200 to 300|1598|
|300 to 400|1383|
|400 to 500|1043|
|500 to 600|688|
|600 to 700|252|
|700 to 800|98|
|800 to 900|39|
|900 to 1000|18|
|1000 to 10 000|25|

Not that much difference seen by swopping garbage collectors for the database approach.

**Comment by sageserpent-open** [comment]
Using Curium, Z GC:

2 hours, 23 minutes, 46 seconds --> 1 287 155 requests

Memory used fluctuates continually between 7.5Gb and 11.5 Gb.

CPU usage fluctuates between 5% and 25%.

Batches can take >~ 1 second, but settle down to less than once every 4000 requests (200 batches).


|Milliseconds|Number of Requests|
|---|---|
|1 to 100|1676|
|100 to 200|3469|
|200 to 300|388|
|300 to 400|20|
|400 to 500|80|
|500 to 600|20|
|600 to 700|0|
|700 to 800|20|
|800 to 900|0|
|900 to 1000|0|


Quite an improvement!

**Comment by sageserpent-open** [comment]
Let's revisit with a parallelism of 100, so 100 requests in a batch....

Imperative approach using Doobie, Z GC:

2 hours, 23 minutes, 43 seconds --> 217 096 requests

|Milliseconds|Number of Requests|
|---|---|
|1 to 100|0|
|100 to 1000|181|
|1000 to 2000|4550|
|2000 to 3000|1065|
|3000 to 4000|0|


So stuffing more requests into a batch makes the batch take proportionally longer with no overall improvement over the cases where the batch size was 20.

**Comment by sageserpent-open** [comment]
Using Curium, Z GC:

2 hours, 23 minutes, 18 seconds --> 2 585 299 requests

Batches often take >~ 1 second, but settle down to less than once every 5709 requests (57 batches).

|Milliseconds|Number of Requests|
|---|---|
|1 to 100|0|
|100 to 200|1518|
|200 to 300|2338|
|300 to 400|1230|
|400 to 500|323|
|500 to 600|140|
|600 to 700|60|
|700 to 800|100|
|800 to 900|0|
|900 to 1000|0|

So there is some economy in making the batches larger, although obviously they do tend to take longer than when the batch size is 20.

**Comment by sageserpent-open** [comment]
Visual VM for a second run using Curium and Z GC:

100 requests per batch in this run too.

<img width="1750" alt="Screenshot 2023-04-17 at 15 53 37" src="https://user-images.githubusercontent.com/1765601/232523979-5c431afb-c22a-42a9-8eae-204cdf54255e.png">

**Comment by sageserpent-open** [comment]
Graph of batch durations **taking at least 1 second** against the number of requests completed for the run referenced above, 100 requests per batch, Z GC:

![image](https://user-images.githubusercontent.com/1765601/232525049-539d77cf-f82c-4171-8571-b4ef58d3c386.png)

**Comment by sageserpent-open** [comment]
Variation on the last benchmark run, so again, 100 requests per batch, using the default GC for Java 14.

First Visual VM:

<img width="1764" alt="Screenshot 2023-04-17 at 20 32 36" src="https://user-images.githubusercontent.com/1765601/232591683-f8265e15-6a60-4803-beeb-cbf207ea6dfd.png">

Observe the regular spikes in CPU usage, correlated with the regular sawtooth memory pattern as GC kicks in big waves.

**Comment by sageserpent-open** [comment]
Graph of batch durations taking at least 1 second against the number of requests completed for the run referenced above, 100 requests per batch, default Java 14 GC:

![image](https://user-images.githubusercontent.com/1765601/232593069-c08e7d65-9c04-4248-8d22-f624530af7ab.png)

Note the regular big durations of over 10 seconds for batches to complete, these correspond to the GC waves.

**Comment by sageserpent-open** [comment]
Next up, tidy up the very messy code and transplant the changes from the benchmarking repository to Curium...

**Comment by sageserpent-open** [comment]
How much difference does the use of proxies make in the benchmark?

Disabling proxies leads to 35697 requests being processed in 8 minutes and 3 seconds, with big spikes of 12 to 13 seconds for a batch.

Visual VM:
<img width="1756" alt="Screenshot 2023-04-21 at 09 29 58" src="https://user-images.githubusercontent.com/1765601/233586937-1aa14f68-db31-49b6-84c6-635623eb0394.png">

Enabling proxies leads to 88896 requests being processed in 8 minutes and 2 seconds, spikes are all <= 3 seconds, the overwhelming majority being around 1 to 1.5 seconds.

Visual VM:
<img width="1755" alt="Screenshot 2023-04-21 at 09 33 47" src="https://user-images.githubusercontent.com/1765601/233587653-434e9a31-ace6-4aac-83b6-61b019fb4fff.png">

As there is already evidence that memory is not exhausted over several hours, it is clear that the proxy mechanism works well.

**Comment by sageserpent-open** [comment]
The implementation changes mentioned above and a more streamlined API is now available as of commit SHA: 884e8b74428cf850945e6ce2d85b9c1dde4265bb on the `master` branch.

**Comment by sageserpent-open** [comment]
Not quite .... revisiting the _other_ benchmark, `ImmutableObjectStorageMeetsMap` showed problems with that commit - since then, a whole lot of trial and error has been expended to try and get decent performance with *both* benchmarks.

It is up for debate as to whether commit 884e8b7 actually worked that well for `Benchmark`.

What needs to be established is:

a) What commit yielded the marvellous results for the hacked version? Currently re-running `Benchmark` for commit 6164119aa775cab436e1bacc26fba8d868044b67 (in the private repository), which is definitely hacked, has the right date as the results and seems to have the same performance as the results quoted above for a batch size of 1000...

b) How well did this commit work with `ImmutableObjectStorageMeetsMap`?

c) What is the situation with commit 884e8b7 for both benchmarks?

d) Depending on the outcome of b) and c), when did the performance regressions occur and were there separate performance problems for the two benchmarks?

**Comment by sageserpent-open** [comment]
Commit 6164119aa775cab436e1bacc26fba8d868044b67 in the private repository yields 2 339 199 requests (batch size of 100) in 2 hours, 9 minutes and 7 seconds, so we are probably in the ballpark with that commit. The graphs from Visual VM both in the initial 9 minutes and at the end have the same features for the CPU and memory profiles.

So how does that commit fare with `ImmutableObjectStorageMeetsMap`?

**Comment by sageserpent-open** [comment]
To put things on an even footing, `ImmutableObjectStorageMeetsMap` and the tranche cache has been tweaked for parity between the two implementations - so on both branches, it uses:
1. A batch size of 100.
2. Look-back range over the last 100 000 items stored in the map.
3. Thirty second duration after access for a cached tranche.

Z-GC was used for both runs.

Using the old, hacked implementation, said benchmark processes 2 005 000 steps in 837 573 milliseconds. Memory use fluctuates around a level of 7Gb, spiking up to 13 to 4Gb in regular cycles.

Using the new, cleaned-up implementation, the same benchmark processes a mere 540 000 steps in 896 304 milliseconds. Memory use grows and subsides in in long waves from around 7Gb to 9Gb with regular spikes of an additional 2 to 3Gb.

(Incidentally, using the old, hacked implementation with a tranche cache eviction of 5 minutes, as per commit 6164119aa775cab436e1bacc26fba8d868044b67, also yields a throughput of 2 075 000 steps in 827 087 milliseconds, so the big loss in performance in the new code probably isn't just down to some tweak on the tranches cache).

So clearly there has been some performance regression during the clean up!

**Comment by sageserpent-open** [comment]
~~As an aside, it is clear from running with both the old, hacked code and the new, cleaned-up code that `ImmutableObjectStorageMeetsMap` won't fly unless the look-back range is kept down to 100 000 items. Having it at 10 000 000 results requires more memory than 16Gb, presumably because so many tranches are loaded to cope with all the jumping round in such a large range.~~

I'm not so sure about this now - tried re-running with the old, hacked code and a cached tranche duration of 30 seconds, got 2 305 000 steps in 886 089 milliseconds, memory usage has a baseline of 7.5Gb with regular spikes up to 13.5 to 14.5 Gb.

Let's try again and leave it running for a while....

**Comment by sageserpent-open** [comment]
Revisiting the dog-food benchmark `Benchmark`, using the old, hacked implementation but with a cached tranche duration of 30 seconds still yields good performance: 2 526 698 requests in 2 hours, 22 minutes and 43 seconds. So again, it seems that the precise duration of tranche caching is not the fundamental reason for poor performance.

**Comment by sageserpent-open** [comment]
Running `ImmutableObjectStorageMeetsMap` using the old, hacked implementation, varying the look-back...

Exhibit A - Look-back of 10 000 000; 13 100 000 steps in 8 986 532 milliseconds.
<img width="1761" alt="Screenshot 2023-04-28 at 14 43 26" src="https://user-images.githubusercontent.com/1765601/235298131-f9250a43-1f5a-4c1d-8f07-6f8281abcf3b.png">

Exhibit B - Look-back of 5 000 000; 13 030 000 steps in 9 568 485 milliseconds.
<img width="1762" alt="Screenshot 2023-04-28 at 18 19 54" src="https://user-images.githubusercontent.com/1765601/235298172-994dec00-d082-42cd-a023-8fc6a5b510c0.png">

Exhibit C - Look-back of 100 000; 9 610 000 steps in 5 249 961 milliseconds.
<img width="1765" alt="Screenshot 2023-04-28 at 19 16 20" src="https://user-images.githubusercontent.com/1765601/235298228-3bd314da-540d-4523-8bc9-5a931640b6d8.png">

Exhibit D - Look-back of 100 000 + use of default GC instead of Z GC used in all other runs; 12 535 000 steps in 4 234 997 milliseconds.
<img width="1756" alt="Screenshot 2023-04-28 at 20 27 57" src="https://user-images.githubusercontent.com/1765601/235298309-27c5cbdb-a6a0-4174-bb28-a1792ca1a282.png">

**Comment by sageserpent-open** [comment]
It is clear that regardless of the look-back amount, there is a slow creep of memory underlying the spikes that leads to memory exhaustion - it is at that point that long delays of around 10 to 30 seconds become apparent, presumably due to cache thrashing (the benchmark does keep running and will perform fairly snappily in between these long delays).

For example from exhibit A (B is similar).

```
Step: 12790000, duration to before query: 9028210 milliseconds, query duration: 738 milliseconds, minimum update duration: 123 milliseconds, maximum update duration: 370 milliseconds, average update duration: 187

Step: 12795000, duration to before query: 9040442 milliseconds, query duration: 823 milliseconds, minimum update duration: 133 milliseconds, maximum update duration: 309 milliseconds, average update duration: 229

Step: 12800000, duration to before query: 9063071 milliseconds, query duration: 768 milliseconds, minimum update duration: 113 milliseconds, maximum update duration: 12447 milliseconds, average update duration: 435

Step: 12805000, duration to before query: 9071711 milliseconds, query duration: 157 milliseconds, minimum update duration: 91 milliseconds, maximum update duration: 341 milliseconds, average update duration: 156

Step: 12810000, duration to before query: 9080803 milliseconds, query duration: 386 milliseconds, minimum update duration: 110 milliseconds, maximum update duration: 378 milliseconds, average update duration: 178

Step: 12815000, duration to before query: 9090211 milliseconds, query duration: 696 milliseconds, minimum update duration: 117 milliseconds, maximum update duration: 375 milliseconds, average update duration: 179

Step: 12820000, duration to before query: 9100224 milliseconds, query duration: 790 milliseconds, minimum update duration: 136 milliseconds, maximum update duration: 438 milliseconds, average update duration: 185

Step: 12825000, duration to before query: 9109865 milliseconds, query duration: 735 milliseconds, minimum update duration: 119 milliseconds, maximum update duration: 305 milliseconds, average update duration: 176

Step: 12830000, duration to before query: 9120793 milliseconds, query duration: 700 milliseconds, minimum update duration: 133 milliseconds, maximum update duration: 438 milliseconds, average update duration: 203

Step: 12835000, duration to before query: 9142601 milliseconds, query duration: 1145 milliseconds, minimum update duration: 104 milliseconds, maximum update duration: 13440 milliseconds, average update duration: 421

Step: 12840000, duration to before query: 9151403 milliseconds, query duration: 339 milliseconds, minimum update duration: 92 milliseconds, maximum update duration: 303 milliseconds, average update duration: 152

Step: 12845000, duration to before query: 9160755 milliseconds, query duration: 426 milliseconds, minimum update duration: 129 milliseconds, maximum update duration: 332 milliseconds, average update duration: 179

Step: 12850000, duration to before query: 9169562 milliseconds, query duration: 560 milliseconds, minimum update duration: 118 milliseconds, maximum update duration: 261 milliseconds, average update duration: 167

Step: 12855000, duration to before query: 9178962 milliseconds, query duration: 266 milliseconds, minimum update duration: 116 milliseconds, maximum update duration: 468 milliseconds, average update duration: 176

Step: 12860000, duration to before query: 9188210 milliseconds, query duration: 364 milliseconds, minimum update duration: 119 milliseconds, maximum update duration: 251 milliseconds, average update duration: 179

Step: 12865000, duration to before query: 9196444 milliseconds, query duration: 1187 milliseconds, minimum update duration: 126 milliseconds, maximum update duration: 238 milliseconds, average update duration: 156,

Step: 12870000, duration to before query: 9208073 milliseconds, query duration: 967 milliseconds, minimum update duration: 109 milliseconds, maximum update duration: 1396 milliseconds, average update duration: 208

Step: 12875000, duration to before query: 9217422 milliseconds, query duration: 289 milliseconds, minimum update duration: 115 milliseconds, maximum update duration: 281 milliseconds, average update duration: 167

Step: 12880000, duration to before query: 9228028 milliseconds, query duration: 355 milliseconds, minimum update duration: 119 milliseconds, maximum update duration: 326 milliseconds, average update duration: 205

Step: 12885000, duration to before query: 9238121 milliseconds, query duration: 480 milliseconds, minimum update duration: 131 milliseconds, maximum update duration: 428 milliseconds, average update duration: 194

Step: 12890000, duration to before query: 9247441 milliseconds, query duration: 504 milliseconds, minimum update duration: 129 milliseconds, maximum update duration: 238 milliseconds, average update duration: 176

Step: 12895000, duration to before query: 9256425 milliseconds, query duration: 531 milliseconds, minimum update duration: 130 milliseconds, maximum update duration: 228 milliseconds, average update duration: 169

Step: 12900000, duration to before query: 9266828 milliseconds, query duration: 282 milliseconds

Step: 12905000, duration to before query: 9295694 milliseconds, query duration: 688 milliseconds, minimum update duration: 155 milliseconds, maximum update duration: 18486 milliseconds, average update duration: 571

Step: 12910000, duration to before query: 9304192 milliseconds, query duration: 363 milliseconds, minimum update duration: 105 milliseconds, maximum update duration: 224 milliseconds, average update duration: 155

Step: 12915000, duration to before query: 9312820 milliseconds, query duration: 346 milliseconds, minimum update duration: 123 milliseconds, maximum update duration: 227 milliseconds, average update duration: 164

Step: 12920000, duration to before query: 9321827 milliseconds, query duration: 826 milliseconds, minimum update duration: 129 milliseconds, maximum update duration: 282 milliseconds, average update duration: 172

Step: 12925000, duration to before query: 9331395 milliseconds, query duration: 601 milliseconds, minimum update duration: 106 milliseconds, maximum update duration: 298 milliseconds, average update duration: 174

Step: 12930000, duration to before query: 9340068 milliseconds, query duration: 476 milliseconds, minimum update duration: 122 milliseconds, maximum update duration: 326 milliseconds, average update duration: 160

Step: 12935000, duration to before query: 9348932 milliseconds, query duration: 931 milliseconds, minimum update duration: 119 milliseconds, maximum update duration: 252 milliseconds, average update duration: 167

Step: 12940000, duration to before query: 9361196 milliseconds, query duration: 1087 milliseconds, minimum update duration: 116 milliseconds, maximum update duration: 2476 milliseconds, average update duration: 226
)
Step: 12945000, duration to before query: 9370802 milliseconds, query duration: 305 milliseconds, minimum update duration: 107 milliseconds, maximum update duration: 327 milliseconds, average update duration: 169

Step: 12950000, duration to before query: 9383314 milliseconds, query duration: 846 milliseconds, minimum update duration: 103 milliseconds, maximum update duration: 1878 milliseconds, average update duration: 243

Step: 12955000, duration to before query: 9392959 milliseconds, query duration: 832 milliseconds, minimum update duration: 136 milliseconds, maximum update duration: 267 milliseconds, average update duration: 175

Step: 12960000, duration to before query: 9403428 milliseconds, query duration: 226 milliseconds, minimum update duration: 120 milliseconds, maximum update duration: 417 milliseconds, average update duration: 192

Step: 12965000, duration to before query: 9410902 milliseconds, query duration: 294 milliseconds, minimum update duration: 110 milliseconds, maximum update duration: 287 milliseconds, average update duration: 144

Step: 12970000, duration to before query: 9421067 milliseconds, query duration: 388 milliseconds, minimum update duration: 149 milliseconds, maximum update duration: 284 milliseconds, average update duration: 196

Step: 12975000, duration to before query: 9445316 milliseconds, query duration: 889 milliseconds, minimum update duration: 141 milliseconds, maximum update duration: 14609 milliseconds, average update duration: 476

Step: 12980000, duration to before query: 9454535 milliseconds, query duration: 432 milliseconds, minimum update duration: 120 milliseconds, maximum update duration: 344 milliseconds, average update duration: 166

Step: 12985000, duration to before query: 9461846 milliseconds, query duration: 270 milliseconds, minimum update duration: 91 milliseconds, maximum update duration: 266 milliseconds, average update duration: 137
```

**Comment by sageserpent-open** [comment]
The obvious culprit here is the top level map in the benchmark code - it simply accumulates lots of entries, even though updates are limited by the look-back to change entries within a limited range. Although in theory the use of proxies should allow the older entries to vanish, there is still a need to have a proxy somewhere to stand in for the purged state, and that has to sit in the map. The map itself also has proxies, but these are intermixed with large arrays that are not released from memory.

So possibly the map is just getting bigger and bigger.

What happens if the `ImmutableObjectStorageMeetsMap` is modified to remove older entries from the map once the map reaches the size of the look-back limit (so the map throws away entries that are no longer eligible for modification)?

**Comment by sageserpent-open** [comment]
In the meantime while the latest run is being executed, observe the transition in the CPU and memory profiles, this is common to all the runs seen so far with the old code:
<img width="1756" alt="Screenshot 2023-04-29 at 11 03 10" src="https://user-images.githubusercontent.com/1765601/235299016-6d20e116-2193-47b8-b15b-a2fdca39cb7f.png">

**Comment by sageserpent-open** [comment]
After modifying `ImmutableObjectStorageMeetsMap` to purge ageing entries so as to keep its size within the look-back limit, we have gloriously flattened CPU and memory profiles - vindication!

<img width="1753" alt="Screenshot 2023-04-29 at 12 44 20" src="https://user-images.githubusercontent.com/1765601/235300836-00e66c92-0b60-48ef-ba0d-4567842e82e4.png">

So, 13 475 000 steps in 11 062 689 milliseconds, using a batch size of 100 and a tranche cache duration of 30 seconds.

Looking at the performance sometime after the map reached the look-back size limit and thus performance stabilised, we have:

```
Average query time: 368 milliseconds
Average minimum in 50 batches batch update time: 127 milliseconds
Average maximum in 50 batches batch update time: 608 milliseconds
Average over average in 50 batches batch update time: 210 milliseconds
```

Batches are of 100 steps with successive batches separate by a standalone query, every 5000 steps new statistics are collated to give a minimum, maximum and average over the 50 batches, eg:

```
Step: 13405000, duration to before query: 10772216 milliseconds, query duration: 493 milliseconds, minimum update duration: 151 milliseconds, maximum update duration: 518 milliseconds, average update duration: 209

Step: 13410000, duration to before query: 10782954 milliseconds, query duration: 164 milliseconds, minimum update duration: 150 milliseconds, maximum update duration: 423 milliseconds, average update duration: 204

Step: 13415000, duration to before query: 10794694 milliseconds, query duration: 207 milliseconds, minimum update duration: 150 milliseconds, maximum update duration: 990 milliseconds, average update duration: 231

```

So each step takes on average 210 / 100 = 2.1 milliseconds, which isn't bad. Maximum latency on a batch is below 1 second, so that is workable.

**Comment by sageserpent-open** [comment]
Recapitulating some of the changes in the branch that has suffered a performance regression...

**Comment by sageserpent-open** [comment]
Leaving no stone unturned, re-executing `Benchmark` using the tranche cache duration of 30 seconds still yields decent performance: 47 minutes 59 seconds --> 782 598 requests.
<img width="1762" alt="Screenshot 2023-04-30 at 09 10 44" src="https://user-images.githubusercontent.com/1765601/235342815-a4f66da4-30fc-4e98-90fa-60eeb70f04c4.png">

So going with a 30 second retention for both benchmarks from now on...

**Comment by sageserpent-open** [comment]
Change: remove caching of reference id to object look-up entries in `IntersessionState`.

Re-executing `Benchmark`: 2 hours, 26 minutes, 43 seconds ---> 3 505 196 requests

Compare with results from above with look-up included:

2 hours, 23 minutes, 18 seconds --> 2 585 299 requests (5 minutes duration)
2 hours, 9 minutes and 7 seconds --> 2 339 199 requests (5 minutes duration)
2 hours, 22 minutes and 43 seconds --> 2 526 698 requests (also 30 seconds duration)

It seems that this is a slight optimisation, and that benefit was also observed on the other branch too.

As for `ImmutableObjectStorageMeetsMap`, it yields 15 165 000 steps in 15 984 752 milliseconds.

To do an apples-to-apples comparison, the same run yields 13 475 000 steps in 10 860 668 milliseconds, compare with the result above of 13 475 000 steps in 11 062 689 milliseconds. So again, a slight optimisation.

So that change is now in.

<img width="1753" alt="Screenshot 2023-04-30 at 09 26 04" src="https://user-images.githubusercontent.com/1765601/235343394-cca09fd8-5d34-4d31-b895-677c0d967015.png">

**Comment by sageserpent-open** [comment]
Cherry picked changes from the branch that lead to poor performance - currently committed 3994e78eebda1c1b0938a29797cd9961423ce77a which breaks out `ImmutableObjectStorageImplementation` and makes `ImmutableObjectStorage` into a cleaner abstraction (also includes earlier work on restructuring the API so that immutable state and the tranches implementation are encapsulated, as well as the breakout of the configuration).

Results look OK, `ImmutableObjectStorageMeetsMap` yields 13 850 000 steps in 12 440 340 milliseconds.

To do an apples-to-apples comparison, the same run yields 13 475 000 steps in 11 374 876 milliseconds, compare with the previous results of 13 475 000 steps in 10 860 668 milliseconds and 13 475 000 steps in 11 062 689 milliseconds. Some performance loss there, though ...

<img width="1754" alt="Screenshot 2023-05-01 at 12 52 22" src="https://user-images.githubusercontent.com/1765601/235448126-6f0afdbc-c9ea-4ee2-b3f3-a39dfbcf2a0f.png">

**Comment by sageserpent-open** [comment]
Commit d0442e16ee73cb0e00cc8975ca09ad216c4c8a93 - encapsulation of the tranche cache in `IntersessionState` and excision of `SerializationSupport` leads to `ImmutableObjectStorageMeetsMap` yielding 14 785 000 steps in 13 553 349 milliseconds.

To do an apples-to-apples comparison, the same run yields 13 475 000 steps in 10 008 699 milliseconds which is a decent result.

<img width="1745" alt="Screenshot 2023-05-01 at 17 19 19" src="https://user-images.githubusercontent.com/1765601/235486378-c4b3a2d6-5f3a-4b31-a4ed-d55128b34cf3.png">

**Comment by sageserpent-open** [comment]
Commit 3e972c1e0a8050adca3a5850c5d7faa88afad1cf - proxy support broken out, use of configuration to set up the tranches cache leads to `ImmutableObjectStorageMeetsMap` yielding 14 105 000 steps in 14 830 197 milliseconds.

To do an apples-to-apples comparison, the same run yields 13 475 000 steps in 12 890 002 milliseconds. This isn't so good.

What about `Benchmark`?

**Comment by sageserpent-open** [comment]
While `Benchmark` is running, review of the cherry pick shows that the excision of proxy support was mixed in with some other code changes intended to simplify translation from canonical object reference ids to tranche local ones. Possibly that had caused the slight performance regression?

There was also a subtle change in `ImmutableObjectStorageMeetsMap` in the commit that encapsulated the intersession state (and use of tranches) - previously there were two instances of `ImmutableObjectStorage`, one for querying and the other for updating. Perhaps this has had an impact, it certainly was the case that introducing the split had a performance benefit....

**Comment by sageserpent-open** [comment]
`Benchmark`: 49 minutes 11 seconds --> 821 959 requests. Not much difference there. Let's break down the recent cherry-pick by partial reversion and see the effects...

**Comment by sageserpent-open** [comment]
Commit 22ce8b179d3256c5a3346295949ed75e6f4bb46a reinstates the use of `StandaloneObjectLookup`, but sticks with `ProxySupport` being excised.

This leads to `ImmutableObjectStorageMeetsMap` yielding 14 270 000 steps in 13 247 622 milliseconds.

To do an apples-to-apples comparison, the same run yields 13 475 000 steps in 11 443 499  milliseconds.

The timings seen so far at 13 475 000 steps:

11 374 876 milliseconds
10 860 668 milliseconds
11 062 689 milliseconds
10 008 699 milliseconds
12 890 002 milliseconds
11 443 499  milliseconds

Perhaps this is simply a case of random variation?

Oddly enough, the query duration is rarely reaches as high as 1 millisecond - that didn't used to be the case. When did that start happening?

Possibly connected with using a single instance of `ImmutableObjectStorage` for both querying and updating?

**Comment by sageserpent-open** [comment]
So, before reworking the API (which lead to the same instance of `ImmutableObjectStorage` being used in `ImmutableObjectStorageMeetsMap`), we see:

```
Step: 100000, duration to before query: 12263 milliseconds, query duration: 14 milliseconds, minimum update duration: 13 milliseconds, maximum update duration: 16 milliseconds, average update duration: 14

Step: 105000, duration to before query: 13255 milliseconds, query duration: 18 milliseconds, minimum update duration: 14 milliseconds, maximum update duration: 57 milliseconds, average update duration: 19

Step: 110000, duration to before query: 14152 milliseconds, query duration: 12 milliseconds, minimum update duration: 11 milliseconds, maximum update duration: 60 milliseconds, average update duration: 17

Step: 115000, duration to before query: 14829 milliseconds, query duration: 12 milliseconds, minimum update duration: 11 milliseconds, maximum update duration: 22 milliseconds, average update duration: 12

Step: 120000, duration to before query: 15518 milliseconds, query duration: 11 milliseconds, minimum update duration: 11 milliseconds, maximum update duration: 34 milliseconds, average update duration: 13

Step: 125000, duration to before query: 16168 milliseconds, query duration: 11 milliseconds, minimum update duration: 11 milliseconds, maximum update duration: 14 milliseconds, average update duration: 12

Step: 130000, duration to before query: 16839 milliseconds, query duration: 22 milliseconds, minimum update duration: 11 milliseconds, maximum update duration: 16 milliseconds, average update duration: 12
```

After reworking the API:

```
Step: 100000, duration to before query: 10948 milliseconds, query duration: 0 milliseconds, minimum update duration: 13 milliseconds, maximum update duration: 15 milliseconds, average update duration: 14

Step: 105000, duration to before query: 11675 milliseconds, query duration: 0 milliseconds, minimum update duration: 13 milliseconds, maximum update duration: 15 milliseconds, average update duration: 14

Step: 110000, duration to before query: 12476 milliseconds, query duration: 0 milliseconds, minimum update duration: 13 milliseconds, maximum update duration: 57 milliseconds, average update duration: 15

Step: 115000, duration to before query: 13466 milliseconds, query duration: 0 milliseconds, minimum update duration: 15 milliseconds, maximum update duration: 39 milliseconds, average update duration: 19

Step: 120000, duration to before query: 14136 milliseconds, query duration: 0 milliseconds, minimum update duration: 10 milliseconds, maximum update duration: 28 milliseconds, average update duration: 12

Step: 125000, duration to before query: 14749 milliseconds, query duration: 0 milliseconds, minimum update duration: 11 milliseconds, maximum update duration: 13 milliseconds, average update duration: 11

Step: 130000, duration to before query: 15364 milliseconds, query duration: 0 milliseconds, minimum update duration: 11 milliseconds, maximum update duration: 14 milliseconds, average update duration: 11
```

I'll take a query duration of less than 1 millisecond as the preferred option!

Assuming a degree of random variation even over quite long runs of `ImmutableObjectStorageMeetsMap`, let's revert the partial reversion and carry on cherry-picking...

**Comment by sageserpent-open** [comment]
Cherry picking and squashing the first two of a series of commits to do with storing objects created in a session in a front-end cache that persists for the duration of a session results in awful performance...

Running `Benchmark` for 3 minutes (running it twice) processes  48198, 51896  requests prior to this commit and 14698, 14299 requests after.

I'll see if further cherry-picking sorts this out, but if not, then let's go with what we've got.

**Comment by sageserpent-open** [comment]
No, further cherry-picking did not resolve the issue - and as the commits done up to the introduction of caching of objects stored in a session pass the additional testing introduced in that work, it seems like a good place to stop.

Let's re-run the benchmarks...

**Comment by sageserpent-open** [comment]
Running `Benchmark` processes 3 545 492 requests in 2 hours 47 minutes 11 seconds.


<img width="883" alt="Screenshot 2023-05-02 at 19 40 06" src="https://user-images.githubusercontent.com/1765601/235756725-0b796102-ffbf-4f9e-9305-015e69ed92a1.png">

**Comment by sageserpent-open** [comment]
Running `ImmutableObjectStorageMeetsMap` yields 12 910 000 requests in 9 022 295 milliseconds.

<img width="1754" alt="Screenshot 2023-05-02 at 22 09 19" src="https://user-images.githubusercontent.com/1765601/235787463-f097c50d-2706-4099-b272-ece6ca3e1a20.png">

**Comment by sageserpent-open** [comment]
Delivered in the Curium repository as of commit SHA: 87aa297008b3a11de71ba81af6eb615c24ea75f2 .

Phew!
